src/global/Pattern.h

```diff
@@ -25,7 +25,11 @@ static const PatternID NO_PATTERN = std::numeric_limits<PatternID>::max();
  *        while writing a query).
  */
 struct Pattern {
-  Id& operator[](const size_t pos) { return _data[pos]; }
+  using value_type = Id;
+  using ref = value_type&;
+  using const_ref = const value_type&;
+
+  value_type& operator[](const size_t pos) { return _data[pos]; }
```

Comment:
Where do you need this and how can the compiler tell whether to use this operator or the one from the next line?

Reply:
&* These are overloads that only differ in the constness of the result AND of the method. The compiler chooses this method if `this` is not const, and else the second one. This is an absolute standard pattern for containers, and the magic behind the following behavior:

```
std::vector<int> a{3};
a[0] = 42; // perfectly fine, the non-const operator[] returns a non-const reference.
const std::vector<int> b{3};
b[0] = 43; // compiler error, b is const, the const overload of operator[] returns a const ref to which we may not write
```


I am not sure if this is needed in one of the other PRs, but in general it is good to have a complete interface (other methods of the `Pattern` class also have the const and non-const overloads, e.g. `back`.